### PR TITLE
Enable monitoring

### LINKED ISSUE #147: Bundle more assets to stackable artifacts
## The problem

Some operators (zk, kafka, spark, nifi) now require a `product config spec` file to validate various configuration properties. This file is located in the git repository under `deploy/config-spec/properties.yaml` but is not included in the generated artefacts.

Without them, these operators cannot start anymore after being deployed from a CI build.

## Proposed solution

Include this file in the artifacts and probably update operator code to load them from the appropriate location.

There is a similar problem with the CRDs. In the planning meeting from 21.07.2021 it was decided to bundle those with the artifacts too.

Implemented in #146

### PR DESCRIPTION
**Pull request by maltesander**
- Added metrics port to CR
- Added ENV variable for javaagent to start jmx exporter

Also fixes #147

### CODE REVIEW

File: deploy/config-spec/properties.yaml [hunk 1]
```diff
@@ -103,4 +103,18 @@ properties:
...
+      description: "Amount of time, in ticks (see `tickTime`), to allow followers to sync with ZooKeeper. If followers fall too far behind a leader, they will be dropped. In other words: The number of ticks that can pass between sending a request and getting an acknowledgment before a follower is dropped."
+
+  - property: &metricsPort
+      propertyNames:
+        - name: "metricsPort"
+          kind:
+            type: "env"
+      datatype:
+        type: "string"
+        unit: *unitPort
```
Comment (razvan) [inline deploy/config-spec/properties.yaml:115]:
```suggestion
        unit: *unitPort
        min: "1024"
        max: "65535"
```

File: deploy/config-spec/properties.yaml [hunk 2]
```diff
@@ -103,4 +103,18 @@ properties:
...
+      roles:
+        - name: "server"
+          required: false
+      asOfVersion: "0.0.0"
+      description: "The port where ZooKeeper metrics are exposed."
```
Comment (razvan) [inline deploy/config-spec/properties.yaml:120]:
```suggestion
      description: "The port where ZooKeeper metrics are exposed as a Prometheus endpoint."
      
```

File: operator/src/lib.rs [hunk 3]
```diff
@@ -748,6 +751,23 @@ impl ZookeeperState {
...
+                        // product config to be able to not configure any monitoring / metrics)
+                        if property_name == "metricsPort" {
+                            if let Some(port) = &property_value {
+                                // cannot panic because checked in product config
+                                metrics_port = Some(port.parse::<u16>()?);
+                                env_vars.push(EnvVar {
+                                name: "SERVER_JVMFLAGS".to_string(),
+                                // TODO: avoid that "{{" and "}}" formatting
+                                value: Some(format!("-javaagent:{}packageroot{}/{}/stackable/lib/jmx_prometheus_javaagent-0.16.1.jar={}:{}packageroot{}/{}/stackable/conf/jmx_exporter.yaml",
+                                            "{{", "}}", version.package_name(), port,  "{{", "}}", version.package_name())),
```
Comment (razvan) [inline operator/src/lib.rs:764]:
```suggestion
                                value: Some(format!("-javaagent:{{{{packageroot}}}}/{}/stackable/lib/jmx_prometheus_javaagent-0.16.1.jar={}:{{{{packageroot}}}}/{}/stackable/conf/jmx_exporter.yaml",
                                             version.package_name(), port,  version.package_name())),
```

**Review by razvan — COMMENTED** [review]
Some suggestions

File: crd/src/lib.rs [hunk 4]
```diff
@@ -50,7 +51,11 @@ impl Configuration for ZookeeperConfig {
         _resource: &Self::Configurable,
         _role_name: &str,
     ) -> Result<BTreeMap<String, Option<String>>, ConfigError> {
-        Ok(BTreeMap::new())
+        let mut result = BTreeMap::new();
+        if let Some(metrics_port) = self.metrics_port {
+            result.insert("metricsPort".to_string(), Some(metrics_port.to_string()));
```
Comment (siegfriedweber) [inline crd/src/lib.rs:56]: Set `SERVER_JVMFLAGS` already here and remove the section from `create_pod_and_config_maps`.

File: operator/src/lib.rs [hunk 5]
```diff
@@ -800,11 +820,17 @@ impl ZookeeperState {
...
+
+        let mut annotations = BTreeMap::new();
+        // only add metrics container port and annotation if available
+        if let Some(metrics_port) = metrics_port {
+            annotations.insert(SHOULD_BE_SCRAPED.to_string(), "true".to_string());
+            container_builder.add_container_port(
+                ContainerPortBuilder::new(metrics_port)
+                    .name("metrics")
+                    .build(),
+            );
```
Comment (siegfriedweber) [inline operator/src/lib.rs:833]: `clientPort` could also be added to the container ports.

**Review by siegfriedweber — CHANGES_REQUESTED** [review]

File: crd/src/lib.rs (hunk 4 above)
Comment (maltesander) [reply crd/src/lib.rs]: Im checking it right now. The problem here is that we would not check that "metricsPort" via the product-config if we add it here. But in general yeah, the `create_pod_and_config_maps` is a monster, it needs to be splitted.

File: operator/src/lib.rs (hunk 5 above)
Comment (maltesander) [reply operator/src/lib.rs:837]: Done.

**Review by siegfriedweber — APPROVED** [review]
LGTM

**Pull request merged**